### PR TITLE
IntersectionObserver API integration

### LINKED ISSUE #139: Fetch links based on Intersection Observer
Mainly based on `quicklink` approach
https://github.com/GoogleChromeLabs/quicklink

- Detects links into the current viewport (using Intersection Observer):
https://github.com/GoogleChromeLabs/quicklink/blob/5e7714b94db1ca702bbb63431b1f9c8f33a04050/src/index.mjs#L22

- Waits until the browser is idle (using requestIdleCallback)
https://github.com/GoogleChromeLabs/quicklink/blob/5e7714b94db1ca702bbb63431b1f9c8f33a04050/src/index.mjs#L85

- Checks if the user isn't on a slow connection (using navigator.connection.effectiveType) or has data-saver enabled (using navigator.connection.saveData)
https://github.com/GoogleChromeLabs/quicklink/blob/5e7714b94db1ca702bbb63431b1f9c8f33a04050/src/prefetch.mjs#L106

- Prefetches URLs to the links (using <link rel=prefetch> or XHR). Provides some control over the request priority (can switch to fetch() if supported).
https://github.com/GoogleChromeLabs/quicklink/blob/5e7714b94db1ca702bbb63431b1f9c8f33a04050/src/index.mjs#L88

Vue.js version: https://github.com/egoist/vue-router-prefetch/blob/master/src/index.js#L7

Related:

- https://github.com/bluebill1049/react-simple-img
- https://developers.google.com/web/updates/2019/02/intersectionobserver-v2

### PR DESCRIPTION
**Pull request by breadadams**
closes #139 

This is a very WIP version of this @Kikobeats, just wanted to get something up here.

I've had to skip husky verification on committing because there seems to be an error with the linter config (?):

```sh
Could not parse lint-staged config.

        Error: ● Validation Error:

Invalid value for 'linters'.

Advanced configuration has been deprecated. For more info, please visit: https://github.com/okonet/lint-staged.

Configured value is: {'*.js': ['prettier-standard', 'git add'], 'package.json': ['finepack', 'git add']}

Please refer to https://github.com/okonet/lint-staged#configuration for more information...
```

### CODE REVIEW

File: packages/react/package.json [hunk 1]
```diff
@@ -34,6 +35,7 @@
     "@storybook/addon-storyshots": "latest",
     "@storybook/react": "latest",
     "babel-eslint": "latest",
+    "babel-loader": "^8.0.6",
```
Comment (breadadams) [inline packages/react/package.json]: Had to add these new dependencies for Storybook

File: packages/react/src/index.js [hunk 2]
```diff
@@ -107,37 +112,58 @@ function Microlink (props) {
...
-          controls={controls}
-          loop={loop}
-          muted={muted}
-          playsInline={playsInline}
-          size={size}
-        />
-      )}
-    </CardWrap>
-  )
+    <div ref={cardRef}>
```
Comment (breadadams) [inline packages/react/src/index.js]: Only temporary, don't worry 😛

File: packages/react/stories/index.js [hunk 3]
```diff
@@ -41,6 +41,9 @@ storiesOf('props', module)
   .add('loading', () =>
     createMicrolink({ url: 'https://microlink.io', loading: true })
   )
+  .add('useIntersection', () =>
+    urls.map(url => createMicrolink({ url, useIntersection: true }))
```
Comment (Kikobeats) [inline packages/react/stories/index.js]: what do you think to use `lazy` as naming and use `true` by default?

Comment (breadadams) [reply packages/react/stories/index.js]: Yeah, the current name is horrible 😅. I'm not sure about the default `true` though, wouldn't it be better for users to opt-in to it?

Comment (breadadams) [reply packages/react/stories/index.js]: I've renamed the prop to `lazy` @Kikobeats but the default is still `false`

File: packages/react/src/components/Card/CardWrap.js [hunk 4]
```diff
@@ -76,12 +76,14 @@ const createEl = el =>
       contrast && (!color || !backgroundColor) && hoverStyle
   )
 
-const CardWrap = ({ rel, href, target, ...props }) => {
+const CardWrap = forwardRef(({ href, rel, target, ...restProps }, ref) => {
```
Comment (Kikobeats) [inline packages/react/src/components/Card/CardWrap.js]: why `forwardRef` is necessary? 🤔

File: packages/react/src/index.js [hunk 5]
```diff
@@ -150,6 +176,7 @@ Microlink.defaultProps = {
   playsInline: true,
   direction: 'ltr',
   size: 'normal',
+  useIntersection: false,
```
Comment (Kikobeats) [inline packages/react/src/index.js]: this is referenced only here, never used

File: packages/react/src/components/Card/CardWrap.js (hunk 4 above)
Comment (breadadams) [reply packages/react/src/components/Card/CardWrap.js]: Good question, so we need to hook onto the outer wrap of the card for the `IntersectionObserver` to listen, and given that this a stateless component React doesn't keep a hold of the ref, so you have to use the `forwardRef` to get access to it.

Without this, if you try to pass a `ref` prop to `CardWrap` the node is always `undefined`.

File: packages/react/src/index.js [hunk 6]
```diff
@@ -38,20 +41,42 @@ function Microlink (props) {
...
   } = props
   const isLoadingUndefined = loadingProp === undefined
+  const [hasIntersected, setHasIntersected] = useState(false)
   const [loadingState, setLoading] = useState(
     isLoadingUndefined ? true : loadingProp
   )
   const [state, setState] = useState({})
   const apiUrl = createApiUrl(props)
 
+  const cardRef = useCallback(node => {
```
Comment (breadadams) [inline packages/react/src/index.js]: @Kikobeats the reason behind using `useCallback` is because we depend on the ref "on render". However, when the component is initialized the ref is undefined - until the `CardWrap` hits the DOM that is.

In the initial version of the PR I was combining the `useRef` hook with `useEffect` to combat this. Listening for when the ref changes and attaching the DOM element to the IntersectionObserver listener _then_. I was _also_ doing this inside a conditional block which is of course a no-go with hooks (they **must** be in the root of the component) Eg.

```js
const cardRef = useRef();

if (canBeLazy) {
  useEffect(() => {
    if (cardRef && cardRef.currrent) {
      // And attaching the card to the observer here
    }
  }, [cardRef.current])
}
```

The first fix would obviously be to move any conditional statements to _inside_ the hook, but this got me thinking if there was a better way to combine the ref & effect hooks. As we don't want any observer-related code to run _at all_ unless the `lazy` prop has been specified.

That lead me to [this](https://medium.com/@teh_builder/ref-objects-inside-useeffect-hooks-eb7c15198780) medium article. You'll see midway in the post an embedded tweet where funnily enough someone's asking D. Abramov about precisely the IntersectionObserver API 😅(thread [here](https://twitter.com/dan_abramov/status/1093497348913803265)).

That post also links to a portion of the React docs where they highlight a _similar_ use-case, you can read that [here](https://reactjs.org/docs/hooks-faq.html?source=post_page-----eb7c15198780----------------------#how-can-i-measure-a-dom-node).

Tldr; The `useCallback` hook allows us to create a memoized function that listens for the `CardWrap`'s ref node and uses it for the `IntersectionObserver` if the `lazy` prop is set.

File: packages/react/src/index.js [hunk 7]
```diff
@@ -115,7 +140,8 @@ function Microlink (props) {
       cardSize={size}
       color={color}
       backgroundColor={backgroundColor}
-      isLoading={isLoading}
+      loading={isLoading}
```
Comment (breadadams) [inline packages/react/src/index.js]: I broke this in rebase by mistake, fixing now 🤦‍♂

**Comment by breadadams** [comment]
I've made some finishing touches and refactored the code into a custom hook 🤓

File: packages/react/src/index.js [hunk 8]
```diff
@@ -170,13 +170,13 @@ Microlink.defaultProps = {
   apiKey: undefined,
   autoPlay: true,
   controls: true,
-  media: ['image', 'logo'],
+  direction: 'ltr',
+  lazy: false,
```
Comment (Kikobeats) [inline packages/react/src/index.js]: let set up it as `true` 🙂

**Comment by Kikobeats** [comment]
@breadadams stories need to reflect any non-default setup, so setup `lazy: true` by default and update to the story to reflect `lazy: false` will cover all the possibilities

File: packages/react/.storybook/config.js [hunk 9]
```diff
@@ -1,16 +1,26 @@
...
+addDecorator(withA11y)
+
 addParameters({
   options: {
     theme: {
       brandTitle: 'Microlink SDK',
-      brandUrl: 'https://github.com/microlinkhq/sdk'
+      brandUrl: 'https://github.com/microlinkhq/sdk',
+      brandImage:
+        'https://user-images.githubusercontent.com/5795227/63266705-cd7b4d80-c290-11e9-8906-ad787636b048.png'
```
Comment (breadadams) [inline packages/react/.storybook/config.js]: @Kikobeats if you want to upload this image somewhere I'll change the URL

File: packages/react/package.json [hunk 10]
```diff
@@ -31,6 +31,7 @@
     "@babel/plugin-transform-react-inline-elements": "latest",
     "@babel/preset-env": "latest",
     "@babel/preset-react": "latest",
+    "@size-limit/preset-small-lib": "^2.1.0",
```
Comment (Kikobeats) [inline packages/react/package.json]: why this change?

File: packages/react/src/index.js [hunk 11]
```diff
@@ -47,11 +51,19 @@ function Microlink (props) {
   const [state, setState] = useState({})
   const apiUrl = createApiUrl(props)
 
+  const isLazyEnabled = isLazySupported && (lazy === true || isObject(lazy))
+  const lazyOptions = isObject(lazy) ? lazy : {}
```
Comment (Kikobeats) [inline packages/react/src/index.js]:
```suggestion
  const lazyOptions = isObject(lazy) ? lazy : undefined
```

I want to be sure we use the `intersectionObserver` default options

File: packages/react/package.json (hunk 10 above)
Comment (breadadams) [reply packages/react/package.json]: In the `@latest` version of the library they've changed the setup to 3 different "types", it appears to be "Web app", "component/UI library" or "general library" - I think this best match is the latter.

You'll spot in Travis that since [this build](https://travis-ci.org/microlinkhq/sdk/builds/570504944) it was failing due to the output they show if you're still using `size-limit` w/ `2.1.0`.

image

I _do_ need to change the version to `latest` though, whoops 😅

File: packages/react/src/index.js (hunk 11 above)
Comment (breadadams) [reply packages/react/src/index.js]: Good point :+1:
